keras_cv/layers/vit_patch_embedding.py

```diff
@@ -0,0 +1,141 @@
+        - https://github.com/keras-team/keras-io/blob/master/examples/vision/image_classification_with_vision_transformer.py
+
+    args:
+        - num_patches: the number of input patches to project
+        - project_dim: the dimensionality of the project_dim
+    Basic usage:
+
+    ```
+    patches = keras_cv.layers.Patching(patch_size)(batch_img)
+
+    project_dim = 1024
+    num_patches = patches.shape[1] # 196
+
+    encoded_patches = keras_cv.layers.PatchEmbedding(num_patches, project_dim)(patches)
+    print(encoded_patches.shape) # (1, 197, 1024)
+    ```
+    """
+
+    def __init__(self, num_patches, project_dim, **kwargs):
+        super().__init__(**kwargs)
+        self.num_patches = num_patches
+        self.project_dim = project_dim
+        self.linear_projection = layers.Dense(self.project_dim)
+        self.position_embedding = layers.Embedding(
+            input_dim=self.num_patches + 1, output_dim=project_dim
+        )
+
+    def call(
+        self,
+        patch,
+        interpolate=False,
+        interpolate_width=None,
+        interpolate_height=None,
+        patch_size=None,
+    ):
+        # Add learnable class token before linear projection and positional embedding
+        patch = ClassTokenizing()(patch)
+        # num_patches + class token
+        positions = tf.range(start=0, limit=self.num_patches + 1, delta=1)
+
+        if interpolate and None not in (
+            interpolate_width,
+            interpolate_height,
+            patch_size,
+        ):
+            encoded = self.linear_projection(
+                patch
+            ) + self.interpolate_positional_embeddings(
+                self.position_embedding(positions),
+                interpolate_width,
+                interpolate_height,
+                patch_size,
+            )
+        elif interpolate and None in (
+            interpolate_width,
+            interpolate_height,
+            patch_size,
+        ):
+            raise ValueError(
+                f"`None of `interpolate_width`, `interpolate_height` and `patch_size` cannot be None if `interpolate` is True"
+            )
+        else:
+            encoded = self.linear_projection(patch) + self.position_embedding(positions)
+        return encoded
+
+    def interpolate_positional_embeddings(self, embedding, height, width, patch_size):
+        """
+        Allows for pre-trained position embedding interpolation. This trick allows you to fine-tune a ViT
+        on higher resolution images than it was trained on.
+        Based on:
+        https://github.com/huggingface/transformers/blob/main/src/transformers/models/vit/modeling_tf_vit.py
+        """
+
+        dimensionality = embedding.shape[-1]
+
+        class_token = tf.expand_dims(embedding[:1, :], 0)
+        patch_positional_embeddings = embedding[1:, :]
+
+        h0 = height // patch_size
+        w0 = width // patch_size
+
+        new_shape = tf.constant(int(math.sqrt(self.num_patches)))
```

Comment:
One thing that hurts is this because of `tf.consant(int(math...))`.
I tried `tf.math`, but even with several `tf.cast()`s, it still had issues with dtypes :/